source/common/network/upstream_server_name.h

```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include "envoy/stream_info/filter_state.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * Server name to set in the upstream connection. The filters like tcp_proxy should use this
+ * value to override the server name specified in the upstream cluster, for example to override
+ * the SNI value in the upstream TLS context.
+ */
+class UpstreamServerName : public StreamInfo::FilterState::Object {
+public:
+  UpstreamServerName(absl::string_view server_name) : server_name_(server_name) {}
+  const std::string& value() const { return server_name_; }
+  static const std::string Key;
```

Comment:
What is this for?

Reply:
This is needed to get an instance of `UpstreamServerName` from the `StreamInfo`'s `FilterState` of the downstream connection: 

https://github.com/envoyproxy/envoy/pull/4973/files/88b3d312e7c28b8ba965b280273e05a50e2ec56d#diff-a5570d8b9053a47a1547728cb6c90989R367

`tcp_proxy` gets this instance from the downstream connection, creates an instance of `TransportSocketOptions` and passes it to `tcpConnPoolForCluster`.

The pattern is similar to `PerConnectionCluster` 
https://github.com/envoyproxy/envoy/blob/0b2e9f5eefc56d57b6fbf38577187c2a036c1208/source/common/tcp_proxy/tcp_proxy.h#L161.

Code in tcp_proxy.cc that processes `PerConnectionCluster ` : https://github.com/envoyproxy/envoy/blob/0b2e9f5eefc56d57b6fbf38577187c2a036c1208/source/common/tcp_proxy/tcp_proxy.cc#L114.

- `StreamInfo::FilterState::Object` is the means to attach data to a downstream connection, by a filter.
- `TransportSocketOptions` is the means to attach data to a transport socket.

Reply:
OK, let me fix it also in tcp_proxy.